### ISSUE TITLE
FileExistsError: [Errno 17] File exists: 'builds'

### ISSUE THREAD

**Issue by ndewijer** [terraform-aws-modules/terraform-aws-lambda#64]
Good day,

I use a setup to build Slack notification lambda's in my environment.

I'm running a container that I set up as follows:

```
FROM hashicorp/terraform:0.12.29

ENV SCRIPTS_DIR=/scripts

RUN mkdir -p ${SCRIPTS_DIR}

WORKDIR ${SCRIPTS_DIR}

COPY *.sh ${SCRIPTS_DIR}/

RUN apk update && \
  apk --no-cache add ruby ruby-bundler ruby-json diffutils jq && \
  # Install awspec
  gem install awspec --no-document && \
  # Install the AWS CLI
  apk --no-cache add python3 py3-pip && \
  ln -sf /usr/bin/python3 /usr/bin/python && \
  pip3 install --upgrade pip && \
  pip3 install awscli && \
  # Install openssl, curl & tflint
  apk --no-cache add openssl curl && \
  wget "$(curl -Ls https://api.github.com/repos/terraform-linters/tflint/releases/latest | grep -o -E "https://.+?_linux_amd64.zip")" && \ 
  unzip tflint_linux_amd64.zip -d /tmp/ && \
  mkdir -p /usr/local/bin && \
  install /tmp/tflint /usr/local/bin/ && \
  # Cleanup
  rm -rf /root/.cache && \
  rm -rf /tmp/* && \
  # chmod scripts
  chmod +x ${SCRIPTS_DIR}/*.sh && \
  ls ${SCRIPTS_DIR}
```

The lambda is in a module I call multiple times.

sns.tf
```
module "notify_slack_ok_cw" {
  source = "git@github.com:**/terraform-aws-excp-notify-slack?ref=v1.0.1"

  providers = {
    aws = aws.cloudwatch
  }
  sns_topic_name       = local.sns_topic_ok_name
  slack_webhook_url    = var.sns_slack_webhook
  slack_channel        = local.sns_slack_channel
  slack_username       = "SNS Alert - OK"
  slack_emoji          = ":tada:"
  lambda_function_name = "${local.lambda_name_ok}-${local.useast1_region}"
  create               = var.create_all
  create_sns_topic     = var.create_sns_topic
  default_tags         = var.default_tags
}

module "notify_slack_urgent_cw" {
  source = "git@github.com:**/terraform-aws-excp-notify-slack?ref=v1.0.1"

  providers = {
    aws = aws.cloudwatch
  }
  sns_topic_name       = local.sns_topic_urgent_name
  slack_webhook_url    = var.sns_slack_webhook
  slack_channel        = local.sns_slack_channel
  slack_username       = "SNS Alert - Urgent"
  slack_emoji          = ":ambulance:"
  lambda_function_name = "${local.lambda_name_urgent}-${local.useast1_region}"
  create               = var.create_all
  create_sns_topic     = var.create_sns_topic
  default_tags         = var.default_tags
}
```

and the lamba call itself looks as follows:

lambda.tf
```

module "lambda" {
  source  = "terraform-aws-modules/lambda/aws"
  version = "1.24.0"

  create = var.create

  function_name = var.lambda_function_name
  description   = "Lambda function to post CloudWatch Event Notifications to a Slack Channel"
  handler       = "notify_slack.lambda_handler"
  runtime       = "python3.6"
  timeout       = 30

  reserved_concurrent_executions = var.reserved_concurrent_executions

  source_path = "${path.module}/functions/notify_slack.py"

  environment_variables = {
    SLACK_WEBHOOK_URL = var.slack_webhook_url
    SLACK_CHANNEL     = var.slack_channel
    SLACK_USERNAME    = var.slack_username
    SLACK_EMOJI       = var.slack_emoji
  }

  tags = var.default_tags
}
```

We use workspaces and var-files in our plan.

When running the terraform plan it sometimes errors out with:

```
Error: failed to execute "python3": Traceback (most recent call last):
  File ".terraform/modules/sns.sns_slack.notify_slack_urgent_cw.lambda/package.py", line 1245, in <module>
    main()
  File ".terraform/modules/sns.sns_slack.notify_slack_urgent_cw.lambda/package.py", line 1241, in main
    exit(args.command(args))
  File ".terraform/modules/sns.sns_slack.notify_slack_urgent_cw.lambda/package.py", line 1085, in prepare_command
    os.makedirs(artifacts_dir)
  File "/usr/lib/python3.8/os.py", line 223, in makedirs
    mkdir(name, mode)
FileExistsError: [Errno 17] File exists: 'builds'


  on .terraform/modules/sns.sns_slack.notify_slack_urgent_cw.lambda/package.tf line 7, in data "external" "archive_prepare":
   7: data "external" "archive_prepare" {
```

The root directory has a folder called build with a few json's in them.

**Comment by garybaines** [comment]
We also get the same error intermittently and are trying to solve it, as it's making using this module in our pipelines a no-go at the moment.  The problem is there's no consistency to generating it.

For background, we're running our terraform plan/apply in an AWS CodeBuild project using the Amazon 'amazonlinux2-x86_64-standard' image
```
269 | Error: failed to execute "python3": Traceback (most recent call last):
270 | File ".terraform/modules/CreateCustodianInfra.notify_slack.lambda/package.py", line 791, in <module>
271 | main()
272 | File ".terraform/modules/CreateCustodianInfra.notify_slack.lambda/package.py", line 787, in main
273 | exit(args.command(args))
274 | File ".terraform/modules/CreateCustodianInfra.notify_slack.lambda/package.py", line 562, in prepare_command
275 | os.makedirs(artifacts_dir)
276 | File "/root/.pyenv/versions/3.7.6/lib/python3.7/os.py", line 221, in makedirs
277 | mkdir(name, mode)
278 | FileExistsError: [Errno 17] File exists: 'builds'
```

**Comment by garybaines** [comment]
Looks to be a race condition on creation of the builds directory, probably due to missing a depends_on in the data.external.archive_prepare resource against the local_file.archive_plan resource (in the package.tf file).  We decided to just do a **_mkdir builds_** before running a plan or apply within our container as a quick work-around.  We're still testing as the error is intermittent, but indicators are that this has solved the problem for us.

**Comment by garybaines** [comment]
Spoke too soon.  Again the error is intermittent, but now we get an error creating the zip file in the builds directory:
```
module.CreateCustodianInfra.module.notify_slack_staging.module.lambda.null_resource.archive[0]: Provisioning with 'local-exec'...
module.CreateCustodianInfra.module.notify_slack_staging.module.lambda.null_resource.archive[0] (local-exec): Executing: ["python3" "../../modules/terraform-aws-lambda/package.py" "build" "--timestamp" "1603203088099219000" "builds/d5713d99898254d5a3e331382ba484242521bb0441dd645d23ebdc4735377476.plan.json"]
module.CreateCustodianInfra.module.notify_slack.module.lambda.null_resource.archive[0] (local-exec): zip: creating 'builds/d5713d99898254d5a3e331382ba484242521bb0441dd645d23ebdc4735377476.zip' archive
module.CreateCustodianInfra.module.notify_slack_staging.module.lambda.null_resource.archive[0] (local-exec): zip: creating 'builds/d5713d99898254d5a3e331382ba484242521bb0441dd645d23ebdc4735377476.zip' archive
module.CreateCustodianInfra.module.notify_slack_staging.module.lambda.null_resource.archive[0] (local-exec): zip: adding: notify_slack.py
module.CreateCustodianInfra.module.notify_slack.module.lambda.null_resource.archive[0] (local-exec): zip: adding: notify_slack.py
module.CreateCustodianInfra.module.notify_slack_staging.module.lambda.null_resource.archive[0] (local-exec): Created: builds/d5713d99898254d5a3e331382ba484242521bb0441dd645d23ebdc4735377476.zip
module.CreateCustodianInfra.module.notify_slack.module.lambda.null_resource.archive[0] (local-exec): Traceback (most recent call last):
module.CreateCustodianInfra.module.notify_slack.module.lambda.null_resource.archive[0] (local-exec):   File "../../modules/terraform-aws-lambda/package.py", line 1245, in <module>
module.CreateCustodianInfra.module.notify_slack.module.lambda.null_resource.archive[0] (local-exec):     main()
module.CreateCustodianInfra.module.notify_slack.module.lambda.null_resource.archive[0] (local-exec):   File "../../modules/terraform-aws-lambda/package.py", line 1241, in main
module.CreateCustodianInfra.module.notify_slack.module.lambda.null_resource.archive[0] (local-exec):     exit(args.command(args))
module.CreateCustodianInfra.module.notify_slack.module.lambda.null_resource.archive[0] (local-exec):   File "../../modules/terraform-aws-lambda/package.py", line 1133, in build_command
module.CreateCustodianInfra.module.notify_slack.module.lambda.null_resource.archive[0] (local-exec):     bpm.execute(build_plan, zs, query)
module.CreateCustodianInfra.module.notify_slack.module.lambda.null_resource.archive[0] (local-exec):   File "../../modules/terraform-aws-lambda/package.py", line 312, in __exit__
module.CreateCustodianInfra.module.notify_slack.module.lambda.null_resource.archive[0] (local-exec):     self.close()
module.CreateCustodianInfra.module.notify_slack.module.lambda.null_resource.archive[0] (local-exec):   File "../../modules/terraform-aws-lambda/package.py", line 302, in close
module.CreateCustodianInfra.module.notify_slack.module.lambda.null_resource.archive[0] (local-exec):     os.replace(self._tmp_filename, self.filename)
module.CreateCustodianInfra.module.notify_slack.module.lambda.null_resource.archive[0] (local-exec): FileNotFoundError: [Errno 2] No such file or directory: 'builds/d5713d99898254d5a3e331382ba484242521bb0441dd645d23ebdc4735377476.zip.tmp' -> 'builds/d5713d99898254d5a3e331382ba484242521bb0441dd645d23ebdc4735377476.zip'

module.CreateCustodianInfra.module.notify_slack_staging.module.lambda.null_resource.archive[0]: Creation complete after 0s [id=7689060501047421484]

Error: Error running command 'builds/d5713d99898254d5a3e331382ba484242521bb0441dd645d23ebdc4735377476.plan.json': exit status 1. Output: zip: creating 'builds/d5713d99898254d5a3e331382ba484242521bb0441dd645d23ebdc4735377476.zip' archive
zip: adding: notify_slack.py
Traceback (most recent call last):
  File "../../modules/terraform-aws-lambda/package.py", line 1245, in <module>
    main()
  File "../../modules/terraform-aws-lambda/package.py", line 1241, in main
    exit(args.command(args))
  File "../../modules/terraform-aws-lambda/package.py", line 1133, in build_command
    bpm.execute(build_plan, zs, query)
  File "../../modules/terraform-aws-lambda/package.py", line 312, in __exit__
    self.close()
  File "../../modules/terraform-aws-lambda/package.py", line 302, in close
    os.replace(self._tmp_filename, self.filename)
FileNotFoundError: [Errno 2] No such file or directory: 'builds/d5713d99898254d5a3e331382ba484242521bb0441dd645d23ebdc4735377476.zip.tmp' -> 'builds/d5713d99898254d5a3e331382ba484242521bb0441dd645d23ebdc4735377476.zip'
```
And in the 'builds' directory we have:
```
$ ls builds/
d5713d99898254d5a3e331382ba484242521bb0441dd645d23ebdc4735377476.plan.json
d5713d99898254d5a3e331382ba484242521bb0441dd645d23ebdc4735377476.zip
```
Re-running afterwards with these files now in existence works.

Something is off in the package.py script, but can't say I can spot it.  I'd recommend taking a look at https://github.com/JayDoubleu/terraform-external-archive-python, which creates a consistent zip file regardless of OS, and solves having to touch an empty file for the zip, and removes all the issues around having a mixed OS base for developers, as the state file remains the same, so would solve the timestamp diff issue too (https://github.com/terraform-aws-modules/terraform-aws-lambda/issues/63).

**Comment by antonbabenko** [comment]
Thanks a lot for opening this issue!

This issue should have been fixed in #65.

Please use [v1.25.0](https://github.com/terraform-aws-modules/terraform-aws-lambda/releases/v1.25.0) which has been just released.
